infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/multiuser/oauth/OpenshiftTokenInitializationFilter.java

```diff
@@ -0,0 +1,141 @@
+import org.eclipse.che.api.user.server.UserManager;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.commons.subject.Subject;
+import org.eclipse.che.commons.subject.SubjectImpl;
+import org.eclipse.che.multiuser.api.authentication.commons.SessionStore;
+import org.eclipse.che.multiuser.api.authentication.commons.filter.MultiUserEnvironmentInitializationFilter;
+import org.eclipse.che.multiuser.api.authentication.commons.token.RequestTokenExtractor;
+import org.eclipse.che.multiuser.api.permission.server.AuthorizedSubject;
+import org.eclipse.che.multiuser.api.permission.server.PermissionChecker;
+import org.eclipse.che.workspace.infrastructure.openshift.OpenShiftClientFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Singleton
+public class OpenshiftTokenInitializationFilter extends MultiUserEnvironmentInitializationFilter {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(OpenshiftTokenInitializationFilter.class);
+
+  private static final List<String> UNAUTHORIZED_ENDPOINT_PATHS =
+      Collections.singletonList("/system/state");
+
+  private final PermissionChecker permissionChecker;
+  private final OpenShiftClientFactory clientFactory;
+
+  private final UserManager userManager;
+
+  @Inject
+  public OpenshiftTokenInitializationFilter(
+      SessionStore sessionStore,
+      RequestTokenExtractor tokenExtractor,
+      OpenShiftClientFactory clientFactory,
+      UserManager userManager,
+      PermissionChecker permissionChecker) {
+    super(sessionStore, tokenExtractor);
+    this.clientFactory = clientFactory;
+    this.userManager = userManager;
+    this.permissionChecker = permissionChecker;
+  }
+
+  @Override
+  protected String getUserId(String token) {
+    try {
+      return getCurrentUser(token).getMetadata().getUid();
+    } catch (KubernetesClientException e) {
+      if (e.getCode() == 401) {
+        LOG.error(
+            "Unauthorized when getting current user. Invalid OpenShift token, probably expired. Re-login? Re-request the token?");
+      }
+      throw new RuntimeException(e);
+    } catch (InfrastructureException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  protected Subject extractSubject(String token) {
+    try {
+      ObjectMeta userMeta = getCurrentUser(token).getMetadata();
+      User user =
+          userManager.getOrCreateUser(
+              userMeta.getUid(), openshiftUserEmail(userMeta), userMeta.getName());
+      return new AuthorizedSubject(
+          new SubjectImpl(user.getName(), user.getId(), token, false), permissionChecker);
+    } catch (InfrastructureException | ServerException | ConflictException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private io.fabric8.openshift.api.model.User getCurrentUser(String token)
+      throws InfrastructureException {
+    OpenShiftClient client = clientFactory.createAuthenticatedClient(token);
```

Comment:
Oh... new client for each request ? :dont_know:

Reply:
Do you have any suggestions how to improve it? My assumption is that creating new client is relatively "cheap" operation. I'm looking how we use it and we usually create a client only for single call and let it die. However, I'm open to new ideas.

Reply:
too heavy refactoring, if we want to make it everyvere.  +0 to left it as is.

Reply:
I agree that somehow caching the clients might be a good idea we're constructing a new okhttp client with a set of interceptors in that call, there are at least two URLs parsed, etc.

But at the same time, let's not optimize prematurely, I think a comment that we're aware of the potential problems would be good here.